### PR TITLE
Tweak Statistics API for implementation

### PR DESCRIPTION
**Pull request by windy1**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1434) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/1076)

This PR modifies the existing Statistics API to comply with my implementation.

#### `StatisticGroup` has been replaced with `StatisticType`
The old system didn't really make a lot of sense for the implementation and was rather arbitrary. `StatisticType` is for special statistics on blocks, items, and entities. All other statistics have the `BASIC` type.

----

#### Builders and API registration has been removed
Registering a statistic or achievement from the API makes about as much sense as doing so for an `ItemType` or `BlockType` so I decided to remove API registration altogether.

----

#### Removed `StatisticFormat`
Instead I just decided to expose the `NumberFormat` the `Statistic` uses.

Signed-off-by: Walker Crouse <walkercrouse@hotmail.com>

### CODE REVIEW

File: src/main/java/org/spongepowered/api/statistic/Statistic.java [hunk 1]
```diff
@@ -25,103 +25,30 @@
...
+     * Returns the objective {@link Criterion} for this statistic.
      *
-     * @return The statistic group this statistic belongs to
+     * @return Objective criterion
      */
-    StatisticGroup getGroup();
+    Optional<Criterion> getCriterion();
 
-    interface Builder extends StatisticBuilder<Statistic, Builder> {
+    NumberFormat getFormat();
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/statistic/Statistic.java:54]: Javadocs?

File: src/main/java/org/spongepowered/api/statistic/Statistics.java [hunk 2]
```diff
@@ -33,101 +33,109 @@
 
     // SORTFIELDS:ON
 
-    public static final Statistic ANIMALS_BRED = DummyObjectProvider.createFor(Statistic.class, "ANIMALS_BRED");
+    public static final Statistic OPEN_SHULKER_BOX = createFor(Statistic.class, "OPEN_SHULKER_BOX");
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/statistic/Statistics.java]: Looks like you need to run the sort fields task.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java [hunk 3]
```diff
@@ -24,34 +24,19 @@
...
-    /**
-     * Creates a new {@link Builder} to build a {@link Achievement}.
-     *
-     * @return The new builder
-     */
-    static Builder builder() {
-        return Sponge.getRegistry().createBuilder(Builder.class);
-    }
-
+public interface Achievement extends Statistic {
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:41]: I originally decided against this as it for a number of reasons. Primarily that the actual statistics value is pretty useless for achievements and representing it as such is a bit of a hack. I'm guessing you made this change for ease of implementation?

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java [hunk 4]
```diff
@@ -75,89 +60,17 @@ static Builder builder() {
...
     /**
-     * If this achievement is backed by a statistic (eg. if
-     * {@link #getSourceStatistic()} is not absent) then this will return the
-     * target value which must be reached for this achievement to be awarded to
-     * a player.
+     * Returns true if this is a "special" achievement.
      *
-     * @return The target value, if available
-     */
-    Optional<Long> getStatisticTargetValue();
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:92]: Was this moved anywhere else? I don't see it.

**Review by Deamon5550 — CHANGES_REQUESTED** [review]
I can live with not having the registration although I'm not sure why you choose to remove it?

**Comment by liach** [comment]
@Deamon5550 When I commented on the commit several hours ago, the pull was not yet opened. Will comment here.

File: src/main/java/org/spongepowered/api/data/key/Keys.java [hunk 5]
```diff
@@ -59,6 +59,7 @@
 import org.spongepowered.api.entity.living.monster.Skeleton;
 import org.spongepowered.api.entity.living.player.Player;
 import org.spongepowered.api.entity.projectile.arrow.Arrow;
+import org.spongepowered.api.statistic.Statistic;
```
Comment (liach) [inline src/main/java/org/spongepowered/api/data/key/Keys.java:62]: Import order

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java [hunk 6]
```diff
@@ -75,89 +60,17 @@ static Builder builder() {
...
 
     /**
-     * Gets the {@link Statistic} which this achievement is backed by if it is available.
+     * Returns the {@link ItemStack} that this achievement is represented by.
      *
-     * @return The source statistic, if available
+     * @return ItemStack representation
      */
-    Optional<Statistic> getSourceStatistic();
+    Optional<ItemStack> getItemStack();
```
Comment (liach) [inline src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java]: I suggest the use of `ItemStackSnapshot` here.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 3 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:41]: Essentially yes. However, I did implement it in such a way that they are always separated by the API. `StatData` will not contain `Achievements` and `AchievementData` will not contain `Statistics`. I think this is fine since there is no actual value in a `Statistic` and all the other implemented methods still make sense for an `Achievement`.

**Comment by windy1** [comment]
Just going to post what I said to @Deamon5550 in IRC for context.

I chose to remove API registration because it makes about as much sense as allowing API registration for say ItemType or BlockType. Statistics and Achievements require a client side mod to be a "full statistic" and while it would be possible to register stats only on the server, this would cause the server to start sending junk stat IDs to the client.

I do understand the possible use case of server-only stats but I came to the conclusion that it wasn't worth it since a plugin could easily simulate their own stats. Statistics are just an Object->Int map anyways.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 3 above)
Comment (Deamon5550) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:41]: Yeah that sound fine.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 4 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:92]: As far as I could tell there is no relationship between achievements and "source" statistics? I could be wrong though. Also an achievements value is 0 for locked and >0 for unlocked although this is not exposed by the API.

File: src/main/java/org/spongepowered/api/statistic/Statistics.java [hunk 7]
```diff
@@ -24,7 +24,7 @@
  */
 package org.spongepowered.api.statistic;
 
-import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+import static org.spongepowered.api.util.generator.dummy.DummyObjectProvider.createFor;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/statistic/Statistics.java]: Don't static import the method `createFor` please. Yes, it does add to line length, but it's clearer when reading where it's coming from.

File: src/main/java/org/spongepowered/api/statistic/Statistic.java (hunk 1 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/statistic/Statistic.java:54]: How would this formatter be used for the time format?

Expected result: 5d 4h...

File: src/main/java/org/spongepowered/api/statistic/Statistics.java (hunk 7 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/statistic/Statistics.java]: IMO its quite obvious where this method comes from (especially since its a CatalogType list)

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 3 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:41]: If its impl side connected i don't care, but having this connected from the API side does not make sense IMO.

>  StatData will not contain Achievements and AchievementData will not contain Statistics. 

This sentences alone is a reason to separate it API wise. (They don't have anything in common)

The format method and statistic type do not make sense for Achievements at all.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java [hunk 8]
```diff
@@ -75,89 +60,17 @@ static Builder builder() {
...
-        Builder targetValue(long value);
-
-        /**
-         * Builds and registers an instance of an {@link Achievement}.
-         *
-         * @return A new instance of a achievement
-         * @throws IllegalStateException If the achievement is not completed
-         */
-        Achievement buildAndRegister() throws IllegalStateException;
+    boolean isSpecial();
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:75]: What is a "special" achievement? Is this pertaining the color of the achievement?

File: src/main/java/org/spongepowered/api/statistic/Statistic.java [hunk 9]
```diff
@@ -25,103 +25,30 @@
...
 
+    default StatisticType getType() {
+        return StatisticTypes.BASIC;
     }
 
-    /**
-     * Represents a builder to create new and custom instances of {@link Statistic}
-     * s.
-     */
-    interface StatisticBuilder<T extends Statistic, B extends StatisticBuilder<T, B>> extends ResettableBuilder<T, B> {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/statistic/Statistic.java:78]: Why did you remove the builders?

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/statistic/Statistic.java (hunk 1 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/statistic/Statistic.java:54]: Implement it yourself. The format really has nothing to do with the server so providing an implementation to "mock" the client is entirely arbitrary.

File: src/main/java/org/spongepowered/api/statistic/Statistic.java (hunk 9 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/statistic/Statistic.java:78]: I've already explained this in the OP and again in the comments.

File: src/main/java/org/spongepowered/api/CatalogTypes.java [hunk 10]
```diff
@@ -253,7 +252,6 @@
     public static final Class<HorseStyle> HORSE_STYLE = HorseStyle.class;
 
     @Deprecated public static final Class<HorseVariant> HORSE_VARIANT = HorseVariant.class;
-
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/CatalogTypes.java:256]: Shouldn't be deleting the empty line between these.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 8 above)
Comment (Deamon5550) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:75]: Its displayed as a change of the border around them. example

File: src/main/java/org/spongepowered/api/statistic/BlockStatistic.java [hunk 11]
```diff
@@ -25,35 +25,19 @@
...
-        /**
-         * Sets the {@link BlockType} of this {@link BlockStatistic}.
-         *
-         * @param block The block
-         * @return This builder, for chaining
-         */
-        Builder block(BlockType block);
-
+    default BlockType getBlockType() {
+        return getItemType().getBlock().get();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/statistic/BlockStatistic.java:40]: I'm guessing this is always guaranteed? If not, in any case, I'd recommend doing an `orElseThrow` with an appropriate exception based on the `ItemType`, perhaps something to do with the implementation.

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 6 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java]: Likewise, yes, I'd recommend using `ItemStackSnapshot` to avoid having to constantly copy the `ItemStack` defensively when returning it.

**Review by gabizou — CHANGES_REQUESTED** [review]
Minor changes, good job so far :)

File: src/main/java/org/spongepowered/api/CatalogTypes.java (hunk 10 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/CatalogTypes.java:256]: Huh weird, I think `sortClassFields` might have done this for some reason. Maybe because of the annotation?

File: src/main/java/org/spongepowered/api/statistic/BlockStatistic.java (hunk 11 above)
Comment (windy1) [reply src/main/java/org/spongepowered/api/statistic/BlockStatistic.java:40]: Yes the implementing class `SpongeBlockStatistic` will always have a valid `BlockType`.

Comment (gabizou) [reply src/main/java/org/spongepowered/api/statistic/BlockStatistic.java:40]: Gotcha.

File: src/main/java/org/spongepowered/api/CatalogTypes.java (hunk 10 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/api/CatalogTypes.java:256]: Perhaps. Might just need to manually add it and re-run to test.

Comment (windy1) [reply src/main/java/org/spongepowered/api/CatalogTypes.java:256]: Can confirm the gradle task was causing this.

**Comment by windy1** [comment]
@gabizou Added and tested your requested changes on API and Common PR

File: src/main/java/org/spongepowered/api/statistic/Statistic.java (hunk 1 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/statistic/Statistic.java:54]: Then you can also drop this NumberFormat for the same reasons.
Especially since its not very usable for all instances (time + distance).

File: src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java (hunk 8 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java:75]: Isn't this a client only thing? Or is there something special for the server as well? If not then this should probably be dropped until we add client support. (to avoid API breakage)

**Review by ST-DDT — APPROVED** [review]

**Comment by windy1** [comment]
Anything else before these get merged?

**Review by gabizou — APPROVED** [review]
Good job.
